### PR TITLE
Add suggestions view model

### PR DESCRIPTION
**Pull request by arturs105**

### CODE REVIEW

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+namespace Toggl.Foundation.MvvmCross.ViewModels
+{
+    public sealed class SuggestionsViewModel : BaseViewModel
+    {
+        private readonly ISuggestionProvider[] suggestionProviders;
+
+        public ObservableCollection<ITimeEntry> Suggestions { get; }
+            = new ObservableCollection<ITimeEntry>();
+
+        public SuggestionsViewModel(params ISuggestionProvider[] suggestionProviders)
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: Include an MvvmCross specific interface that's easy to inject; I don't know how the IoC system is going to deal with params + array parameters. In theory it shouldn't be a problem, but I don't know if registering and array is a good idea.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+            this.suggestionProviders = suggestionProviders;
+        }
+
+        public async override Task Initialize()
+        {
+            await base.Initialize();
+
+            foreach (var provider in suggestionProviders)
+                Suggestions.Add(await provider.GetSuggestion());
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: This sort of defeats the purposes of observables. Awaiting only returns the last one. Some providers will want to return multiple suggestions (and some won't return anything). Also, some might even error. This does not cover any of those cases.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs (hunk 1 above)
Comment (arturs105) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: So something like this?

```
interface ISuggestionProviderContainer
{
    List<ISuggestionProvider> Providers { get; }
}
```

Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: Precisely

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+            [Fact]
+            public async Task WorksIfProviderHasMultipleSuggestions()
+            {
+                var provider = Substitute.For<ISuggestionProvider>();
+                var timeEntries = new[]
+                {
+                    new TimeEntry{ Description = "te1" },
+                    new TimeEntry{ Description = "te2" },
+                    new TimeEntry{ Description = "te3" }
```
Comment (amulware) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: I'd like spaces before the `{`.

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                };
+                var observable = Observable.Create((IObserver<ITimeEntry> arg) =>
```
Comment (amulware) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: `arg` could be named better.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+using Toggl.Multivac.Models;
+
+namespace Toggl.Foundation.MvvmCross.ViewModels
+{
+    public sealed class SuggestionsViewModel : BaseViewModel
+    {
+        private readonly ISuggestionProviderContainer suggestionProviders;
+
+        public ObservableCollection<ITimeEntry> Suggestions { get; }
+            = new ObservableCollection<ITimeEntry>();
```
Comment (amulware) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: What is the reasoning behind using this interface, if we don't actually need all the properties in it? Would it not make sense to have an interface/model with exactly the properties needed?

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            {
+                Action tryingToConstructWithEmptyParameters =
+                    () => new SuggestionsViewModel(null);
+
+                tryingToConstructWithEmptyParameters
+                    .ShouldThrow<ArgumentNullException>();
+            }
+        }
+
+        public class TheSuggestionsProperty
```
Comment (amulware) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: I think this should also test that it throws if there are no/null providers, and the case where none of the given providers return any suggestions.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs (hunk 5 above)
Comment (arturs105) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: Do you mean `ITimeEntry`? I didn't think it was necessary to create a new interface for this. We could also use tuples.

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs (hunk 6 above)
Comment (arturs105) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SuggestionsViewModelTests.cs]: The view model can't even be created without providers. `SuggestionProviderContainer` throws an exception if it gets created with no/null providers. That behaviour is already tested in `SugestionProviderContainerTests`.

Your second point is valid though.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs (hunk 5 above)
Comment (amulware) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: This would be a big tuple, so better to have a separate model, if we would change it. What do you think @willsb?

**Comment by amulware** [comment]
Apart from the question about whether to use a separate model or not, I'm fine with this.

If we added a separate model, the same could be used for continuation of other time entries. Just a thought.

File: Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs (hunk 5 above)
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SuggestionsViewModel.cs]: We will have a PR to change the ITimeEntries in the TImeEntriesLogViewModel to a new model. I think the same model could be used here, no? But that's a different PR
